File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/bfs/Problem1600.java
package com.keunwon.algorithm.bfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Problem1600 {
    static int K, W, H;
    static int[][] map;

    final static int[][] moves = { {-1, 0}, {1, 0}, {0, -1}, {0, 1} };
    final static int[][] hoursMoves = {
            {-2, 1}, {-1, 2},
            {1, 2}, {2, 1},
            {2, 1}, {2, -1},
            {1, -2}, {-1, -2},
            {-2, -1}, {-2, -1}};

    public static void main(String[] args) throws IOException {
        input();
        bfs();
    }

    static void bfs() {
        Queue<Monkey> queue = new LinkedList<>();
        boolean[][][] visited = new boolean[K + 1][H][W];

        queue.add(new Monkey(0, 0, 0, 0));

        while (!queue.isEmpty()) {
            Monkey m = queue.poll();

            if (!validMap(m.x, m.y)) { continue; }
            if (map[m.x][m.y] == 1) { continue; }
            if (visited[m.skill][m.x][m.y]) { continue; }

            if (m.x == H - 1 && m.y == W - 1) {
                System.out.println(m.distance);
                return;
            }

            visited[m.skill][m.x][m.y] = true;

            for (int[] move : moves) {
                int nx = m.x + move[0];
                int ny = m.y + move[1];

                queue.add(new Monkey(nx, ny, m.distance + 1, m.skill));
            }

            if (m.skill == K) { continue; }

            for (int[] hoursMove : hoursMoves) {
                int nx = m.x + hoursMove[0];
                int ny = m.y + hoursMove[1];

                queue.add(new Monkey(nx, ny, m.distance + 1, m.skill + 1));
            }
        }
        System.out.println(-1);
    }

    static boolean isGo(int x, int y, int skill,  boolean[][][] visited) {
        return map[x][y] == 0 && !visited[skill][x][y];
    }

    static boolean validMap(int x, int y) {
        return x >= 0 && x < H && y >= 0 && y < W;
    }

    static void input() throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        K = Integer.parseInt(br.readLine());

        StringTokenizer st = new StringTokenizer(br.readLine());
        W = Integer.parseInt(st.nextToken());
        H = Integer.parseInt(st.nextToken());
        map = new int[H][W];

        for (int i = 0; i < H; i++) {
            st = new StringTokenizer(br.readLine());

            for (int j = 0; j < W; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }
    }

    static class Monkey {
        int x;
        int y;
        int distance;
        int skill;

        public Monkey(int x, int y, int distance, int skill) {
            this.x = x;
            this.y = y;
            this.distance = distance;
            this.skill = skill;
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem1600.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * Title: 말이 되고픈 원숭이
 * Level: 골드-3
 **/
class Problem1600 {
    private val moves = arrayOf(-1 to 0, 0 to 1, 1 to 0, 0 to -1)
    private val diagonals = arrayOf(-2 to -1, -2 to 1, -1 to -2, -1 to 2, 1 to -2, 1 to 2, 2 to -1, 2 to 1)
    private lateinit var map: Array<IntArray>

    fun solution(k: Int, map: Array<IntArray>): Int {
        this.map = map
        return bfs(k)
    }

    private fun bfs(k: Int): Int {
        val queue = LinkedList<Node>()
        val visited = Array(map.size) { Array(map[0].size) { BooleanArray(k + 1) } }

        queue.offer(Node(0, 0, 0, k))
        visited[0][0][k] = true

        while (queue.isNotEmpty()) {
            val cur = queue.poll()

            if (cur.r == map.lastIndex && cur.c == map[0].lastIndex) return cur.distance

            for ((mr, mc) in moves) {
                val nr = cur.r + mr
                val nc = cur.c + mc

                if (nr !in map.indices || nc !in map[0].indices) continue
                if (visited[nr][nc][cur.skill] || map[nr][nc] == 1) continue

                queue.offer(Node(nr, nc, cur.distance + 1, cur.skill))
                visited[nr][nc][cur.skill] = true
            }

            if (cur.skill == 0) continue

            for ((mr, mc) in diagonals) {
                val nr = cur.r + mr
                val nc = cur.c + mc
                val skill = cur.skill - 1

                if (nr !in map.indices || nc !in map[0].indices) continue
                if (visited[nr][nc][skill] || map[nr][nc] == 1) continue

                queue.offer(Node(nr, nc, cur.distance + 1, skill))
                visited[nr][nc][skill] = true
            }
        }
        return -1
    }

    private data class Node(val r: Int, val c: Int, val distance: Int, val skill: Int)
}

fun main() = System.`in`.bufferedReader().use { br ->
    System.out.bufferedWriter().use { bw ->
        val k = br.readLine().toInt()
        val (w, h) = br.readLine().split(" ").map { it.toInt() }
        val map = Array(h) {
            val st = StringTokenizer(br.readLine())
            IntArray(w) { st.nextToken().toInt() }
        }

        Problem1600().solution(k, map).also { bw.write("$it") }
        bw.flush()
    }
}
